File: IDAUtil/SAP/TCodes/WE05.cs
using lib;
using Microsoft.VisualBasic;
using System;

namespace IDAUtil {
    public class WE05 {
        private readonly ISAPLib sapLib;
        private readonly IWinUtil winUtil;

        public WE05(ISAPLib sapLib, IWinUtil winUtil) {
            this.sapLib = sapLib;
            this.winUtil = winUtil;
        }

        public enum ResponseWE05 {
            noItems,
            successSingleValue,
            successTable
        }

        public ResponseWE05 extractReport(string messageVariant, string folderPath) {

            // 2 - inbound idocs only
            // 51 - Application not posted (Hardstop workflow)
            executeWE05(2.ToString(), 51.ToString(), messageVariant);
            string screenShotPath;
            string fileName;
            screenShotPath = folderPath + @"\" + DateAndTime.Now.ToFileNameFormat() + " WE05.jpg";
            fileName = DateAndTime.Now.ToFileNameFormat() + " WE05.xlsx";
            
            if (sapLib.isPopUp()) {
                sapLib.printScreenOfCurrentSession(screenShotPath);
                sapLib.pressEnter();
                return ResponseWE05.noItems;
            }

            if (isTableExists()) {
                sapLib.openExport();
                sapLib.exportExcel(folderPath, fileName);
                sapLib.printScreenOfCurrentSession(screenShotPath);
                return ResponseWE05.successTable;
            } else {
                screenShotPath = folderPath + @"\" + DateAndTime.Now.ToFileNameFormat() + " single item only found WE05.jpg";
                try {
                    sapLib.printScreenOfCurrentSession(screenShotPath);
                } catch (Exception) {
                    System.Threading.Thread.Sleep(3000);
                    sapLib.printScreenOfCurrentSession(screenShotPath);
                }

                return ResponseWE05.successSingleValue;
            }
        }

        private void executeWE05(string direction, string status, string messageVariant) {
            string dateFrom = DateAndTime.DateAdd(DateInterval.Day, -90, DateAndTime.Now).ToSAPFormat();
            string dateTo = DateAndTime.Now.ToSAPFormat();
            sapLib.enterTCode("WE05");
            sapLib.setText(lib.SAP_ALL_ID.WE05.MESSAGE_VARIAN_FLD, messageVariant);
            if ((messageVariant ?? "") == "FR") {
                sapLib.setMultipleSelection(new[] { "BE" }, lib.SAP_ALL_ID.WE05.MESSAGE_VARIANT_MULTIPLE_SELECTION_BTN);
            }

            sapLib.setText(lib.SAP_ALL_ID.WE05.CREATED_ON_FROM_FLD, dateFrom);
            sapLib.setText(lib.SAP_ALL_ID.WE05.CREATED_ON_TO_FLD, dateTo);

            sapLib.setText(lib.SAP_ALL_ID.WE05.DIRECTION_FLD, direction);
            sapLib.setText(lib.SAP_ALL_ID.WE05.CURRENT_STATUS_FLD, status);

            //filter table by orders only
            sapLib.setText("wnd[0]/usr/tabsTABSTRIP_IDOCTABBL/tabpSOS_TAB/ssub%_SUBSCREEN_IDOCTABBL:RSEIDOC2:1100/ctxtIDOCTP-LOW", "ORDERS04");

            sapLib.pressF8();
        }

        private bool setMessageTypeFilter(string msgType) {
            (sapLib.findById(SAP_ID.OPEN_EXPORT_CONTEXT_MENU_BTN_ARR[0]) as dynamic).selectColumn("MESTYP"); // Message Type
            (sapLib.findById(SAP_ID.OPEN_EXPORT_CONTEXT_MENU_BTN_ARR[0]) as dynamic).pressToolbarButton("&MB_FILTER");
            sapLib.setText(SAP_ID.TABLE_FILTER_FIRST_VALUE_FLD, msgType);
            sapLib.pressEnter();
            return true;
        }

        private bool isTableExists() {
            return sapLib.idExists(lib.SAP_ALL_ID.WE05.SELECTED_IDOCS_TABLE);
        }

        private bool isTableEmpty() {
            return (sapLib.findById(lib.SAP_ALL_ID.WE05.SELECTED_IDOCS_TABLE) as dynamic).RowCount == 0;
        }
    }
}

File: IDAUtil/SAP/TCodes/WE05.vb
Imports [lib]

Public Class WE05

    Private ReadOnly sapLib As ISAPLib
    Private ReadOnly winUtil As IWinUtil

    Public Sub New(sapLib As ISAPLib, winUtil As IWinUtil)
        Me.sapLib = sapLib
        Me.winUtil = winUtil
    End Sub

    Public Enum ResponseWE05
        noItems
        successSingleValue
        successTable
    End Enum

    Public Function extractReport(messageVariant As String, folderPath As String) As ResponseWE05

        '2 - inbound idocs only
        '51 - Application not posted (Hardstop workflow)
        executeWE05(2, 51, messageVariant)

        Dim screenShotPath As String
        Dim fileName As String
        screenShotPath = folderPath & "\" & Now().ToFileNameFormat & " WE05.jpg"
        fileName = Now().ToFileNameFormat & " WE05.xlsx"

        If sapLib.isPopUp() Then
            executeWE05("", "", messageVariant)
            If sapLib.isPopUp() Then
                sapLib.printScreenOfCurrentSession(screenShotPath)
                Return ResponseWE05.noItems
            End If
        End If

        If isTableExists() Then
            sapLib.findById(SAP_ALL_ID.WE05.IDOCS_NODE_TREE).selectedNode("IDoc")

            setMessageTypeFilter("ORDERS")
            'if no orders in the list then export at least sth
            If isTableEmpty() Then
                setMessageTypeFilter("")
            End If

            sapLib.openExport()
            sapLib.exportExcel(folderPath, fileName)
            sapLib.printScreenOfCurrentSession(screenShotPath)
            Return ResponseWE05.successTable
        Else
            screenShotPath = folderPath & "\" & Now().ToFileNameFormat & " single item only found WE05.jpg"
            Try
                sapLib.printScreenOfCurrentSession(screenShotPath)
            Catch ex As Exception
                Threading.Thread.Sleep(3000)
                sapLib.printScreenOfCurrentSession(screenShotPath)
            End Try

            Return ResponseWE05.successSingleValue
        End If

    End Function

    Private Sub executeWE05(direction As String, status As String, messageVariant As String)
        Dim dateFrom As String = DateAdd(DateInterval.Day, -90, Now()).ToSAPFormat
        Dim dateTo As String = Now().ToSAPFormat

        sapLib.enterTCode("WE05")

        sapLib.setText(SAP_ALL_ID.WE05.MESSAGE_VARIAN_FLD, messageVariant)

        If messageVariant = "FR" Then
            sapLib.setMultipleSelection({"BE"}, SAP_ALL_ID.WE05.MESSAGE_VARIANT_MULTIPLE_SELECTION_BTN)
        End If

        sapLib.setText(SAP_ALL_ID.WE05.CREATED_ON_FROM_FLD, dateFrom)
        sapLib.setText(SAP_ALL_ID.WE05.CREATED_ON_TO_FLD, dateTo)

        'sapLib.setText(SAP_ALL_ID.WE05.IDOC_NUMBER_FLD, "185197867")

        sapLib.setText(SAP_ALL_ID.WE05.DIRECTION_FLD, direction)
        sapLib.setText(SAP_ALL_ID.WE05.CURRENT_STATUS_FLD, status)
        sapLib.pressF8()

    End Sub

    Private Function setMessageTypeFilter(msgType As String) As Boolean
        sapLib.findById(SAP_ID.OPEN_EXPORT_CONTEXT_MENU_BTN_ARR(0)).selectColumn("MESTYP") ' Message Type
        sapLib.findById(SAP_ID.OPEN_EXPORT_CONTEXT_MENU_BTN_ARR(0)).pressToolbarButton("&MB_FILTER")
        sapLib.setText(SAP_ID.TABLE_FILTER_FIRST_VALUE_FLD, msgType)
        sapLib.pressEnter()
        Return True
    End Function

    Private Function isTableExists() As Boolean
        Return sapLib.idExists(SAP_ALL_ID.WE05.SELECTED_IDOCS_TABLE)
    End Function

    Private Function isTableEmpty() As Boolean
        Return sapLib.findById(SAP_ALL_ID.WE05.SELECTED_IDOCS_TABLE).RowCount = 0
    End Function


End Class
